Gatsby library search page

jastack opened the pull request:
Added a draft of the plugin search with Algolia. Check it out by going to "/plugins/". The left side search bar functions in a very similar way to the regular sidebar, so I made some changes to the template index.js as well. Let me know if this is a good approach / going in the right direction! Doesn't have filter right now, but should be easy to add. @calcsam

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,104 @@
+          style={{
+            color: `white`,
+            border: `0`,
+            boxShadow: `0 0 0 0`,
```
boxShadow: `none` should work

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,104 @@
+        </Link>
+        <div css={{
+          fontFamily: typography.options.bodyFontFamily.join(`,`),
+          fontSize: `${13 / 16 * 100}%`,
```
scale(n)

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,104 @@
+        <div css={{
+          fontFamily: typography.options.bodyFontFamily.join(`,`),
+          fontSize: `${13 / 16 * 100}%`,
+          lineHeight: `1.1`,
```
rhythm(n)

calcsam replied:
(theoretically a whole # or simple fraction like 3/4)

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,104 @@
+          fontFamily: typography.options.bodyFontFamily.join(`,`),
+          fontSize: `${13 / 16 * 100}%`,
+          lineHeight: `1.1`,
+          paddingTop: `7px`,
```
rhythm(n) 

(theoretically a whole # or simple fraction like 3/4)

calcsam commented on www/src/html.js:
```diff
@@ -72,6 +72,10 @@ export default class HTML extends React.Component {
+          <link
+            rel="stylesheet"
+            href="https://unpkg.com/react-instantsearch-theme-algolia@4.0.0/style.min.css"
+          />
```
do we need / use this?

jastack replied:
I do use their styling for the search box, but mostly included it so I could have something that looked decent fairly quickly. Can definitely take this out and do our own.

calcsam replied:
Got it. Didn't realize we were using it. Let's talk about styling in one fell swoop, but for now just add `rel="preload"`

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,104 @@
+      borderBottom: `1px solid grey`,
+    }}>
+      <div css={{
+        padding: `10px`,
```
use rhythm

KyleAMathews commented:
Looking good! 

https://deploy-preview-3456--gatsbyjs.netlify.com/plugins/

![screen shot 2018-01-09 at 4 17 21 pm](https://user-images.githubusercontent.com/71047/34749845-9cf99aa6-f558-11e7-8c5a-ea11dbe0be7c.png)

One quick bit of feedback — we use Yarn to manage dependencies not NPM so please remove the package-lock.json files that were added.

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,104 @@
+  )
+}
+
+function Result({ hit }){
```
const Result = ({ hit }) => {

calcsam commented:
Awesome to see this on #3003! Thanks @jastack!

Some next steps:
* URL stuff: 
(1) plugin links should be /plugins/name not /packages/name 
(2) search queries should be /plugins?text=URL_ENCODED_SEARCH_STRING so that if a user refreshes they get the same page. As you type things in the search bar the URL should update.

* Right sidebar default is funky. Whenever the search bar is empty this should display the default panel as shown in @rdela's mock (but omit the learn link for now). If there are no results it should be a blank panel

![image](https://user-images.githubusercontent.com/4388699/34750339-76c72f08-f55b-11e7-91d0-9c8751fcf698.png)

* Don't worry about the landing page in the mock -- let's just stick with the search for now.

* We'll want to pull in all plugins, not just official plugins

* Mobile support

Random design stuff:
* The selected plugin should be highlighted
* The left panel should have the light purple background
* Border or box shadow on top, left and bottom.
* whole package result box should be a click target, not just the name
* Margin between the title and the body is too small

![image](https://user-images.githubusercontent.com/4388699/34750497-3204b010-f55c-11e7-8039-108b3d7c71c3.png)

KyleAMathews commented:
>plugin links should be /plugins/name not /packages/name

Is there a compelling reason to change the existing URL structure?

KyleAMathews commented:
On pulling in all plugins — we should include in the search any package on NPM with the `gatsby-plugin` tag. We'll probably have to create a blacklist at some point if the tag gets spammed.

calcsam commented:
> Is there a compelling reason to change the existing URL structure?

@KyleAMathews could be a bit weird if the landing page is /plugins but links go to /packages/x

calcsam commented:
another thought on routing -- since we're not searching strictly plugins, we're searching all packages, let's have this live at /packages (we could make /plugins redirect to /packages too if we want).

calcsam commented:
re docsearch & algolia -- two things:

(1) The medium-term plan is to, at build time, have a source plugin that uses Algolia npm-search (https://blog.algolia.com/yarn-search-javascript-packages/) to pull all package details including the readme into Gatsby data, build a page for each package.

(2) In the short run, this is a great start and probably no need to fiddle further with algolia stuffbefore merging

jastack commented:
Thanks guys! I'm going to work on some fixes based on your feedback then get back to you

shannonbux commented:
@jastack This is exciting! Another thing we can use Algiola for (from my understanding) is pulling more Github data beyond just the READme file. We've gotten feedback that popularity (maybe stars?) and number of downloads could be useful data points. I'm hoping those could appear on the left hand side somehow.

KyleAMathews commented:
>Another thing we can use Algiola for (from my understanding) is pulling more Github data beyond just the READme file. We've gotten feedback that popularity (maybe stars?) and number of downloads could be useful data points.

💯 to this

>This mock shows the title "Gatsby Library" in the search bar

yeah, let's use Gatsby Library

rdela commented:
Also please do not forget "search by Algolia" @jastack https://github.com/gatsbyjs/gatsby/issues/3003#issuecomment-350197330

Probably want to link "Algolia" to either
https://github.com/algolia/npm-search
or
https://www.algolia.com/

like:
Search by [Algolia](https://www.algolia.com/)

The main website
<https://www.algolia.com/>

…is probably safest bet since I think we plan to include both doc search *and* npm-search. @KyleAMathews? @calcsam? thoughts? Should we loop Haroen in to help decide?

jastack commented:
Alright, finally pushed another draft! Check it out at "/packages" now.

@calcsam and @KyleAMathews - the biggest change is that now the search pulls from Algolia's npm-search, instead of from only source packages, which means any package with the "gatsby-plugin" tag shows up. I realize this means more work because as Sam mentioned, we will need to have a source plugin in that pulls in all the other package details to build a page for each. So if it makes sense to stick with source package search in the short-term we can do that.

@shannonbux let me know what you think about the styling. I included more info on each search result, like tags, last updated, and downloads and can add icons too.

KyleAMathews commented:
Let's just do the source plugin for `npm-search` so we can build out pages for all the packages.

Two tweaks to the search

* Remove deprecated packages (if that data is exposed)
* also search for `gatsby-component` keyword (e.g. https://github.com/gatsbyjs/gatsby/pull/3507)

Really starting to come together! Excited to see this happen!

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,184 @@
+  onSearchStateChange = searchState => {
+    clearTimeout(this.debouncedSetState);
+    this.debouncedSetState = setTimeout(() => {
+      this.props.history.replace(`/packages?=${searchState.query}`)
```
keep this for now, but...

...we shouldn't be manually manipulating props // the global , we should be calling navigateTo which is how gatsby handles navigation..... 

....however, this may not work at the moment with search queries due to a separate bug: https://github.com/gatsbyjs/gatsby/issues/2074. (We need to check for differences in location.pathname and location.search)  So hold off on changing this for a bit, I'll let you know when I address that and then you can use navigateTo.

KyleAMathews commented on www/src/pages/packages.js:
```diff
@@ -0,0 +1,55 @@
+              textAlign: `center`,
+            }}
+
+            >Welcome to the Gatsby Plugin and Starter Library!</h1>
```
Let's change "plugin and starter" to just "package" since we're not going to search for starters here

KyleAMathews commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,184 @@
+        justifyContent: `center`,
+      }}>
+      <SearchBox
+        translations={{placeholder: 'Search Gatsby Library by keyword'}}
```
Should be `Search Gatsby Library`

calcsam commented on www/src/components/withUrlSync.js:
```diff
@@ -0,0 +1,91 @@
+  };
+};
+
+const originalPathName = location.pathname;
```
this is going to be evaluated outside of the component lifecycle, which is bad. try evaluating it in a constructor

calcsam replied:
ignore if not using file

calcsam commented on www/src/components/withUrlSync.js:
```diff
@@ -0,0 +1,91 @@
+import React, { Component } from 'react';
```
File should be capitalized 'WithUrlSync' since it's a component

calcsam replied:
.......wait, are we using this file at all? I don't see it referenced anywhere

calcsam commented on www/src/components/withUrlSync.js:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    onSearchStateChange = searchState => {
+      clearTimeout(this.debouncedSetState);
```
see above stuff re debouncing

calcsam replied:
ignore if not using file

calcsam commented on www/src/components/withUrlSync.js:
```diff
@@ -0,0 +1,91 @@
+
+const updateAfter = 700;
+const searchStateToQueryString = searchState => ({
+  q: searchState.query,
```
ignore if not using file

calcsam commented on www/src/components/withUrlSync.js:
```diff
@@ -0,0 +1,91 @@
+    ...(searchState.refinementList.keywords && {
+      keywords: searchState.refinementList.keywords,
+    }),
+  }),
```
ignore if not using file

calcsam commented on www/src/components/withUrlSync.js:
```diff
@@ -0,0 +1,91 @@
+    page: p || 1,
+    refinementList: {
+      ...(keywords && { keywords }),
+      ...(owner && { 'owner.name': owner }),
```
ignore if not using file

calcsam commented on www/src/components/withUrlSync.js:
```diff
@@ -0,0 +1,91 @@
+      this.state = {
+        searchState: queryStringToSearchState(location.search.slice(1)),
+      };
+      window.addEventListener('popstate', ({ state: searchState }) => {
```
ignore if not using file

rdela commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,184 @@
+        <h3 css={{
+          fontSize: rhythm(.75),
+          textAlign: `center`,
+        }}>Search by <Link to={`https://www.algolia.com/`} style={{color: `#744C9E`, border: `none`, boxShadow: `none`}}>Algolia</Link></h3>
```
This looks great, but `Link to` should be `a href` here because external link, currently appends to URL at https://deploy-preview-3456--gatsbyjs.netlify.com/packages/

KyleAMathews replied:
Yeah, should be `<a href="https://www.algolia.com"...`

calcsam commented:
This is super cool! Yeah let's def keep going on this one!

Left are: 
(1) source plugin 
(2) building package pages based on the source plugin
(3) code comments

then it's basically mergeable. mobile support would be very good as well, but prob not a blocker to merge.

shannonbux commented:
@jastack Hi James! It's looking great! The main feedback I have is two-fold:
1. It'd be useful to have icons for # of downloads. Should I send you some? In my prototype I just used standard ones from material design. There might be better ones.
2. I think it's important to fit more plugins in the left-hand scrolling area, because right now only about two of them fit there, and many people will be trying to compare a few plugins against each other. _Anything we can do to make the individual plugin cards smaller is ideal_:
- for example, removing the "last contributed" part might be necessary, or shrinking it by just putting the tiny photo somewhere and introducing a hover state so it expands to show more information. 
- Also, the tags could be removed from that card and placed in some sort of header for the full view on the right hand side instead.
- The space between the search bar and the first plugin should probably be just twice the size of the space above the search bar.
- The space underneath "search powered by Algolia" should be equal to the space above it, and the font probably could be smaller (unless we really need to emphasize it for some reason).
3. I think the selected cards on the left would be more readable if the title is in black with the little box around it being white.

fk commented:
@jastack Great work on this! While I catch up on #3003 and the history behind the implementation, some quick thoughts and questions adding up to Shannon's last comment:

* The plugin list claims to show "155 results", but the list seems a little shorter? ;-) Maybe I'm missing something? If that's an "algolia thing", maybe we could try to not display a "result count" initially.
* Conciously thinking about displaying # of downloads the first time made me wonder if we should also link to the npm plugin page in addition to "Browse source code for this package on GitHub" (on the plugin page itself, not in the list). Maybe use the count + an icon to do so to save space while reducing "Browse source code for this package on GitHub". On the card, we should right-align the download count to allow vertical scanning. Really like that you added that metric!
* In general: Are the "gatsby" and "gatsby-plugin" tags needed at all? Am I right in guessing every plugin will (should) be tagged with those two (or at least one of them)? For 👍  to  removing the tags from the card and placing them in the full view – also as they do not currently have any functionality.
* I'm a bit confused about the "last contributed" part: Do we actually really show who contributed last here? At least for the https://www.gatsbyjs.org/docs/plugins/#official-plugins @KyleAMathews always seems to be the last contributor. Maybe because he's the one merging PRs? Right now I don't get much out of that information.
* > The space underneath "search powered by Algolia" should be equal to the space above it, and the font probably could be smaller (unless we really need to emphasize it for some reason).  
Agreed, we can save a lot of space here. We also should (have to?) probably use their logo.
* I think the plugin name on the card doesn't need to have a background/be inverted – to me it leads to thinking I'd have to click exactly that – the title – to bring up more information about the plugin. Also with the latest commits we now can click the whole card 👍. The grey box also adds visual noise which IMO we don't want in a list. Also, the current gatsbyjs.org doesn't use any similar CTA elements.

shannonbux commented:
Very great comments, Florian and I agree - @jastack, this is great stuff! I
agree that the tags and "last contributor" info might not be that
important. A couple discussion questions:
- instead of having the plugin name with it's own background (I agree, this
makes it look erroneously "clickable"), what do you propose? I suppose the
white cards could just have black text on them, and the purple cards could
have white?
- What do you mean by right aligning the download count on the card? Do you
mean right aligning it *above* the cards or somehow on each card on the
left? Just want to make sure I understand.

On Tue, Jan 16, 2018 at 6:08 PM, Florian Kissling <notifications@github.com>
wrote:

> @jastack <https://github.com/jastack> Great work on this! While I catch
> up on #3003 <https://github.com/gatsbyjs/gatsby/issues/3003> and the
> history behind the implementation, some quick thoughts and questions adding
> up to Shannon's last comment:
>
>    - The plugin list claims to show "155 results", but the list seems a
>    little shorter? ;-) Maybe I'm missing something? If that's an "algolia
>    thing", maybe we could try to not display a "result count" initially.
>    - Conciously thinking about displaying # of downloads the first time
>    made me wonder if we should also link to the npm plugin page in addition to
>    "Browse source code for this package on GitHub" (on the plugin page itself,
>    not in the list). Maybe use the count + an icon to do so to save space
>    while reducing "Browse source code for this package on GitHub". On the
>    card, we should right-align the download count to allow vertical scanning.
>    Really like that you added that metric!
>    - In general: Are the "gatsby" and "gatsby-plugin" tags needed at all?
>    Am I right in guessing every plugin will (should) be tagged with those two
>    (or at least one of them)? For 👍 to removing the tags from the card
>    and placing them in the full view – also as they do not currently have any
>    functionality.
>    - I'm a bit confused about the "last contributed" part: Do we actually
>    really show who contributed last here? At least for the
>    https://www.gatsbyjs.org/docs/plugins/#official-plugins
>    <https://www.gatsbyjs.org/docs/plugins/#official-plugins> @KyleAMathews
>    <https://github.com/kyleamathews> always seems to be the last
>    contributor. Maybe because he's the one merging PRs? Right now I don't get
>    much out of that information.
>    -
>
>    The space underneath "search powered by Algolia" should be equal to
>    the space above it, and the font probably could be smaller (unless we
>    really need to emphasize it for some reason).
>    Agreed, we can save a lot of space here. We also should (have to?)
>    probably use their logo.
>
>    - I think the plugin name on the card doesn't need to have a
>    background/be inverted – to me it leads to thinking I'd have to click
>    exactly that – the title – to bring up more information about the plugin.
>    Also with the latest commits we now can click the whole card 👍. The
>    grey box also adds visual noise which IMO we don't want in a list. Also,
>    the current gatsbyjs.org doesn't use any similar CTA elements.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gatsbyjs/gatsby/pull/3456#issuecomment-358160939>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Ae9o2pqtUSV-VvzwHWSeQm7zJNxPJ_tFks5tLUgEgaJpZM4RYm-i>
> .
>

fk commented:
> I suppose the white cards could just have black text on them, and the purple cards could
have white?

Yes, exactly!

> What do you mean by right aligning the download count on the card?

![image](https://user-images.githubusercontent.com/21834/35092842-e9ff2386-fc40-11e7-984b-8cf2abbd6475.png)

shannonbux commented:
Hey ya'll, and esp. @jastack, just curious how this is coming along and want to figure out what I can do to help!

jastack commented:
Thanks everyone for all the feedback!!! I've finally got another draft up and I think we're getting pretty close. 

@shannonbux @fk here are a few of the changes:

- I moved last contributed and tags from the search cards to the full view page. Do you have a vision of how best to display this metadata? Right now it just sits at the top. Florian - I can def add a link to the npm package page as well as the github (and remove "gatsby-plugin" and "gatsby-component" from the tags). I can also add more -- check out the [yarn display page for gatsby-link ](https://yarnpkg.com/en/package/gatsby-link) to see any data we could add. Shannon, if you have icons I should use definitely send them along! I found the download arrow from a simple search, but am open to anything. 
- I removed some spacing from the search results cards to fit more in -- can revert if you think it would look better.
- I made the Algolia link a little smaller but can change to their logo if we think we need that. They have a prebuilt component with their logo they say is required if you are on a free or community plan. We are essentially using Yarn's Algolia implementation, so I'm not exactly sure what we should do here.
- Made the search box a little wider and removed some spacing between search box and results box.
- Made results number only show when something in search box

@KyleAMathews @calcsam the source plugin for grabbing npm package info mostly works. A few notes:

- There are about 5 packages that have readmes that link to images and that screws up the markdown transformer and graphql query. Right now those packages are left out and not imported. Any thoughts on the best way to deal with those? I could alter the readme when creating the nodes to just delete any problem lines.
- I wrote an additional template file to create pages from the readmes from npm. I think I can actually just alter the existing package template file to accommodate both sources. All that to say I need to clean up a bit, refactor, and add comments, etc.
- Sam - I changed the debouncing method, let me know if that is better.

rdela commented:
https://deploy-preview-3456--gatsbyjs.netlify.com/packages/

Search by [Algolia](https://www.algolia.com/) credit looks great thanks @jastack, any thoughts @Haroenv?

KyleAMathews commented on www/src/components/searchbar-body.js:
```diff
@@ -46,52 +60,117 @@ function Search(){
-    <div css={{
-      borderBottom: `1px solid grey`,
+    <Link
+      to={`/packages/${hit.name}`}
```
Missing the trailing slash

KyleAMathews commented on www/src/components/searchbar-body.js:
```diff
@@ -46,52 +60,117 @@ function Search(){
+
       <div css={{
-        padding: `10px`,
+        fontSize: rhythm(.6),
```
Should be `1/2`

KyleAMathews commented on www/src/pages/packages.js:
```diff
@@ -0,0 +1,55 @@
+                  marginRight: rhythm(3),
+                  fontSize: rhythm(3/4),
+                }}>
+                Please use the search bar and filters to find starters
```
Remove mention of starters + filters since we won't have starters here and we haven't implemented filters yet

calcsam commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+  })
+
+  data.hits.forEach(hit => {
+    if (hit.readme.includes("![")){
```
put in a comment with explanation you put in the PR

calcsam commented on www/gatsby-node.js:
```diff
@@ -116,6 +140,31 @@ exports.createPages = ({ graphql, boundActionCreators }) => {
+          if (localPackagesArr.includes(edge.node.title)){
+            createPage({
+              path: slug,
+              component: slash(packageTemplate),
```
instead of an if statement at top, could do:

`component: localPackagesArr.includes(edge.node.title) ? slash(packageTemplate) :  slash(remotePackageTemplate)`

calcsam commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+    const modified = new Date(hit.modified)
+    const created = new Date(hit.created)
+    hit.modified = modified
+    hit.created = created
```
`hit.created = new Date(hit.created)` should be fine.

calcsam commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+    // Remove unneeded data
+    delete hit.readme
+    delete hit._highlightResult
+    delete hit.versions
```
is there a reason to delete this data?

Haroenv replied:
If this data isn't needed, you can use `attributesToRetrieve` in the search parameters

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,225 @@
+  //     this.props.history.replace(`/packages?=${searchState.query}`)
+  //   }, updateAfter);]
+  //   this.setState({ searchState })
+  // }
```
new way looks good, go ahead & delete comments

calcsam commented on www/src/templates/template-docs-packages.js:
```diff
@@ -10,6 +11,11 @@ class DocsPackagesTemplate extends React.Component {
+    const lastUpdated = `${distanceInWords(new Date(metaData.modified), new Date())} ago`;
+
+
+    console.log(this.props);
```
remove console.log statement

calcsam commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+      ...hit,
+      deprecated: `${hit.deprecated}`,
+      id: parentId,
+      parent: null,
```
created: new Date(hit.created),
modified: new Date(hit.modified)

in lieu of above

calcsam commented on www/src/templates/template-remote-packages.js:
```diff
@@ -0,0 +1,99 @@
+            __html: page.html
+          }}
+        />
+        {/* <MarkdownPageFooter page={page} /> */}
```
remove

calcsam commented on www/src/templates/template-remote-packages.js:
```diff
@@ -0,0 +1,99 @@
+        </div>
+
+        <div className="metadataHeader">
+          <a href={info.repository ? info.repository.url : "" }>Browse source code for this package on Github</a>
```
should we even show this text if there is no source code?

KyleAMathews commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+const algoliasearch = require(`algoliasearch`)
```
This file should be in a `src` directory so we can transpile the JavaScript for older versions of node.

Let's also call this source plugin `gatsby-source-npm-package-search`. Always good to make titles very explicit.

So to fix this:

1. From the root of the gatsby repo, run `npm run plop`
2. follow the prompts for creating a new package — name it `npm-source-npm-package-search`
3. Copy your files over e.g. gatsby-node.js to src/gatsby-node.js

KyleAMathews commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+
+  const data = await index.search({
+    query: ``,
+    filters: `(keywords:gatsby-plugin OR keywords:gatsby-component)`,
```
Let's make the search parameters required options so it's easy to use this in other types of sites.

calcsam commented on www/src/templates/template-remote-packages.js:
```diff
@@ -0,0 +1,99 @@
+    const deprecated = this.props.data.npmPackage.deprecated === "false" ? false : true
+    const page = this.props.data.npmPackage.readme.childMarkdownRemark
+    const info = this.props.data.npmPackage
+    const lastUpdated = `${distanceInWords(new Date(info.modified), new Date())} ago`;
```
this is confusing, `info` should be named `npmPackage`, and having the other variables is also confusing, just use eg `npmPackage.deprecated` in the component hierarchy below

calcsam commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,225 @@
+         />
+          <span css={{paddingLeft: rhythm(.25), fontSize: rhythm(.5), textTransform: `uppercase`}}>{hit.lastPublisher.name}</span>
+          <span css={{paddingLeft: rhythm(.25), fontSize: rhythm(.5)}}>{lastUpdated}</span>
+      </div> */}
```
remove ^

KyleAMathews commented:
Deploy preview for *using-glamor* failed.

Built with commit 610d83d6997594e6465213498c0cc5621cc19cde

https://app.netlify.com/sites/using-glamor/deploys/5a69055ba6188f3dffe66913

Haroenv commented:
Really awesome! Lots to read up on, I am
On mobile now, but saw you’re using algoliasearch. Did you also check out and compare if the React library we have (React InstantSearch) fits the use case here? 

Regarding using the Algolia index as a source, we can make a gatsby plugin that browses an index (there’s a browse api key I can give you). This will go through all of the possible records which match (more than possible with search, because that has a cost which we limit to 1000 at runtime). You can take a look at what I quickly hacked around with Algolia-sitemap (made for the yarn site originally). 

I’ll take a look tomorrow for any questions you have

calcsam commented on www/src/html.js:
```diff
@@ -71,11 +71,11 @@ export default class HTML extends React.Component {
             rel="mask-icon"
             href={`/safari-pinned-tab.svg`}
-            color={colors.gatsby}
+            color="#5bbad5"
```
?

calcsam commented on www/src/html.js:
```diff
@@ -71,11 +71,11 @@ export default class HTML extends React.Component {
           />
-          <meta name="msapplication-config" content={`/browserconfig.xml`} />
-          <script src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js" />
+
```
?

Haroenv commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+
+  console.log(`Grabbing NPM packages...`)
+
+  const data = await index.search({
```
This should be index.browseAll() or browse, see how it's done with https://github.com/algolia/algolia-sitemap/blob/master/index.js

Haroenv commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+const algoliasearch = require(`algoliasearch`)
+const crypto = require(`crypto`)
+
+var client = algoliasearch(`OFCNCOG2CU`, `f54e21fa3a2a0160595bb058179bfb1e`)
```
If browse is used, the key shouldn't be hardcoded but process.env something

Haroenv commented on www/src/assets/down-arrow-white.svg:
```diff
@@ -0,0 +1,39 @@
+</g>
+<g>
+</g>
+<g>
```
this image seems to have some stray `g` tags

Haroenv commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,225 @@
+          fontSize: rhythm(.55),
+          textAlign: `center`,
+          margin: rhythm(.75),
+        }}>Search by <a href={`https://www.algolia.com/`} style={{color: `#744C9E`, border: `none`, boxShadow: `none`}}>Algolia</a></h3>
```
However you decided, but it's also possible to link to the post to how it's built on the forum: https://discourse.algolia.com/t/2016-algolia-community-gift-yarn-package-search/319/1

rdela replied:
I for one am eternally grateful for all the help so would defer to you and Algolia. Would the forum post be your preference @Haroenv?

Haroenv replied:
Actually the homepage works best, but you can add `/?utm_campaign=poweredby&utm_content=gatsby`, so we can set up a dedicated page for it

Haroenv commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,184 @@
+
+  render(){
+    return(
+      <div>
```
yep, Instantsearch root creates a div, so not necessary to wrap it

Haroenv commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,74 @@
+    query: ``,
+    filters: `(keywords:gatsby-plugin OR keywords:gatsby-component)`,
+    hitsPerPage: 1000,
+  })
```
You can use `attributesToRetrieve` here to restrict which attributes you want

Haroenv commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,225 @@
+          searchState={this.state.searchState}
+          onSearchStateChange={this.onSearchStateChange.bind(this)}
+          >
+          <Search searchState={this.state.searchState.query} />
```
Add `<Configure attributesToRetrieve={[/* only required attributes*/]} />` here to avoid pulling down the readme etc. for every package, it will make it significantly faster

Haroenv commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,225 @@
+import React, { Component } from 'react'
+import { InstantSearch, Hits, SearchBox, Stats, RefinementList, PoweredBy } from 'react-instantsearch/dom'
```
You can try importing from /es/ to see if bundle size is any different with native esmodules vs commonjs, cc @samouss

Haroenv commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,225 @@
+        paddingTop: rhythm(.25),
+        lineHeight: rhythm(.75),
+      }}>
+        {hit.description}
```
use `<Highlight hit={hit} attributeName="description" tagName="mark" />` here so that it gets highlighted

rdela commented:
@jastack saw Haroen from Algolia comment at https://github.com/gatsbyjs/gatsby/pull/3456/commits/5a4b511c376a135a6daa89fe7007f2f6cb781400#r163770719 think this is Algolia-sitemap repo Haroen refers to in the previous comment in case it applies and you have not already seen: https://github.com/algolia/algolia-sitemap

shannonbux commented:
Ok @jastack, here are icons we could use. They are super basic and in the advanced export you can invert the colors or change them as needed (I guess we'll need to use the Gatsby black color, and I'm not sure what we use for white, @fk?):
* [Sort](https://materialdesignicons.com/icon/sort)
* [Filter](https://materialdesignicons.com/icon/checkbox-marked-outline)
* [# of downloads](https://materialdesignicons.com/icon/download)
* [github stars](https://materialdesignicons.com/icon/star)

fk commented:
@shannonbux This is what we currently have in the "Gatsby palette": https://github.com/gatsbyjs/gatsby/blob/9c13f904adc5037bd30122b07a7f18805942a33c/www/src/utils/colors.js#L60-L76

`gray.dark` is our "black", plain `#ffffff` white or one of the `ui` colors should be fine for "bright-on-dark" scenarios.

shannonbux commented:
Also @jastack here is a mockup showing a possibility for a header that floats above each READme file and also how to indicate # Github stars and # downloads on the small cards on the left. I'm also ok if we leave these things out in the first iteration and add it later if it seems like that kind of data is important (probably none of them are important yet, and will become more important as we get more duplicate plugins and the list of plugins grows). I'm not totally in love with the icons that Sketch let me use. I think the links I sent you just prior to this post are better. And as always, I'm a beginner at this so definitely appreciate feedback.

<img width="657" alt="screen shot 2018-01-25 at 2 27 11 pm" src="https://user-images.githubusercontent.com/32467162/35413241-0432c712-01dc-11e8-809d-a398d82c8bf0.png">

Haroenv commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,225 @@
+            fontWeight: `bold`,
+          }}
+          >
+            {hit.name}
```
`<Highlight hit={hit} attributeName="name" tagName="mark" />`

Haroenv commented on www/src/components/searchbar-body.js:
```diff
@@ -0,0 +1,225 @@
+        <img
+          width="20"
+          height="20"
+          src={hit.lastPublisher.avatar}
```
`hit.owner` instead of `lastPublisher`, it's usually more relevant than whoever published last

gillkyle commented:
I made some progress but I'm trying to resolve a couple of these testing/build errors on the most current commit (cb316ca) to the branch without much luck, the tests that I see failing (4 in total) running `yarn test` give errors akin to this:

```
FAIL  packages/gatsby-transformer-sharp/src/__tests__/gatsby-node.js
  ● Test suite failed to run

    dlopen(/Users/kyle/Kyle/GitHub/plugin-gatsby-library/node_modules/sharp/build/Release/sharp.node, 1): Library not loaded: @rpath/libvips-cpp.42.dylib
      Referenced from: /Users/kyle/Kyle/GitHub/plugin-gatsby-library/node_modules/sharp/build/Release/sharp.node
      Reason: image not found
```

This looked similar to errors on #1754, and seem like they could be issues introduced by changes in the yarn.lock with updated dependencies. Does that seem possible?

And the snapshots with errors (4) are little changes like this. I'm not familiar with when you'd need to update a jest snapshot but the changes I see to the code for the plugin library don't look like they'd be causing an issue like this, and it could be the result of something that was merged oddly.

```
- Snapshot
+ Received
    
-  <pre class="custom-js"><code><span class="token comment" spellcheck="true">// Fake</span>
+  <pre class="custom-js"><code><span class="token comment">// Fake</span>
```

I think I'll reset back to find a commit where these tests weren't failing to identify what really is causing them. Any other ideas @KyleAMathews? Maybe it's even simpler than I think.

calcsam commented:
@gillkyle the changes being caused by an odd merge could definitely be an issue. You may want to try a cherry-picking or interactive rebase strategy to isolate and squash the commits that are part of this branch (the ones by jastack) and rebase on top of a fresh master.

KyleAMathews commented:
@gillkyle tests can sometimes fail locally due to no fault of your own — Gatsby has a ton of packages and if a few of them fail to install correctly (seems like a near guarantee...) then some of tests will fail. Pay more attention to TravisCI failing then what's happening locally.

None of those tests are affected by this PR so there's probably nothing that needs fixed.

Resolve the conflicting files and then see if TravisCI passes tests.

KyleAMathews commented:
@gillkyle turned out that a minor release of one our dependencies broke some of the snapshots 🤷‍♂️ Fixed upstream with https://github.com/gatsbyjs/gatsby/pull/3902. Merge in the lastest master and you'll be good.

gillkyle commented:
I was just looking into it, I'll merge it in and try building it on my branch with Travis CI, thanks for the help!

fk commented:
Just chatted about this with @shannonbux a little and remembered that I had drawn a bunch of icons for the mobile navigation in case "Plugins" became its own section:

<img width="317" alt="plugins-icons" src="https://user-images.githubusercontent.com/21834/36557742-68f30e44-1809-11e8-9f00-60e7246f8731.png">

The two above the "magic wand" icon a) pick up the quite common "puzzle pieces" metaphor for "Plugins", b) show data flowing through those plugins (and/or "Transformers"), all within the Gatsby monogram/"G" shape. Not super clear, so probably not a way to go.

Anyways…questions:

* I [see we decided to go with "Gatsby Library"](https://github.com/gatsbyjs/gatsby/pull/3456#issuecomment-356777189) (which I like a lot)—I suppose non of these icons really fit here anymore?
* Will this section be called "Library" in the main navigation?
* Where do we place it in the main navigation—in between "Tutorial" and "Community"?
* As said I like "Library" a lot, but I also find it a little misleading as I also could imagine the "Gatsby Guides" to be living there. "Plugins" doesn't have this issue IMO. Does that make any sense or am I overthinking?

Adding another item to the main navigation is another nudge to re-think the latter in terms of the home page layout… ;-) For now I think the additional "Library" should fit in with that landing page from ~900px browser width.

Since I'd hate to lose the other navigation items when going below that width, maybe https://css-tricks.com/the-priority-navigation-pattern/ is one way to proceed long-term? Or we could revamp the homepage layout… :-(

fk commented:
Will starters be indexed/part of the library at some point too?

shannonbux commented:
@fk I have no attachment to any particular name at this point; plugins is
probably clearer!

And about starters--we considered combining them with plugins, then decided
against it because we wanted a more visual showcase that in my imagination,
could also be a way to share templates. Open to discussion on where to put
an index of starters.

On Thu, Feb 22, 2018 at 2:28 PM, Florian Kissling <notifications@github.com>
wrote:

> Will starters be indexed/part of the library at some point too?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gatsbyjs/gatsby/pull/3456#issuecomment-367827828>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Ae9o2q1-aYBJ5S1TiljFcstnT4duZgabks5tXdvpgaJpZM4RYm-i>
> .
>

KyleAMathews commented:
>Will starters be indexed/part of the library at some point too?

yeah, like Shannon said. The Library is for plugins and Gatsby-specific React components.

KyleAMathews commented:
Though let's close this PR since work is continuing over on @gillkyle's PR
